tls/s2n_encrypted_extensions.c

```diff
@@ -14,13 +14,64 @@
  */
 
 #include "error/s2n_errno.h"
+#include "utils/s2n_safety.h"
+#include "stuffer/s2n_stuffer.h"
+#include "tls/s2n_tls.h"
+
+/**
+  * Specified in https://tools.ietf.org/html/rfc8446#section-4.3.1
+  * 
+  * In all handshakes, the server MUST send the EncryptedExtensions
+  * message immediately after the ServerHello message.  
+  *
+  * The EncryptedExtensions message contains extensions that can be
+  * protected, i.e., any which are not needed to establish the
+  * cryptographic context but which are not associated with individual
+  * certificates. 
+  **/
 
 int s2n_encrypted_extensions_send(struct s2n_connection *conn)
 {
-    S2N_ERROR(S2N_ERR_UNIMPLEMENTED);
+    struct s2n_stuffer *out = &conn->handshake.io;
+
+    /* Calculate size of encrypted extensions. For minimal TLS 1.3, this is 0
+     * as we are sending an empty EE message
+     */
+    uint16_t total_size = 0;
+
+    /* Write length of extensions */
+    GUARD(s2n_stuffer_write_uint16(out, total_size));
+
+    if (total_size == 0) {
+        return 0;
+    }
+
+    /* Write the extensions to the out buffer. For minimal TLS 1.3, this is
+     * a noop, as we are sending an empty EE message
+     */
+    
+    return 0;
 }
 
 int s2n_encrypted_extensions_recv(struct s2n_connection *conn)
 {
-    S2N_ERROR(S2N_ERR_UNIMPLEMENTED);
+    struct s2n_stuffer *in = &conn->handshake.io;
+    uint16_t extensions_size;
+
+    /* Read encrypted extensions size */
+    S2N_ERROR_IF(2 > s2n_stuffer_data_available(in), S2N_ERR_BAD_MESSAGE);
+    GUARD(s2n_stuffer_read_uint16(in, &extensions_size));
+    S2N_ERROR_IF(extensions_size > s2n_stuffer_data_available(in), S2N_ERR_BAD_MESSAGE);
+
+    /* Process extensions */
+    if (extensions_size > 0) {
+        struct s2n_blob extensions = {0};
+        extensions.size = extensions_size;
+        extensions.data = s2n_stuffer_raw_read(in, extensions.size);
+        notnull_check(extensions.data);
+
+        GUARD(s2n_server_extensions_recv(conn, &extensions));
```

Comment:
I think we should not reuse `s2n_server_extensions_recv()` here as the [spec is clear](https://tools.ietf.org/html/rfc8446#page-37) about the specific extensions allowed in EE.

For a minimal EE implementation, I would prefer we either 
a) drop this line or 
b) duplicate `s2n_server_extensions_recv()` with only the extensions allowed in EE

Ideally, I would love to see all extensions factored out, then have flags or different list/map that specifies which parts of the handshake allows them.

Reply:
I will add a check